File: Ones and Zeroes.py
class Solution(object):
     def findMaxForm(self, strs, m, n):
         """
         :type strs: List[str]
         :type m: int
         :type n: int
         :rtype: int
         """
         s={(0,m,n)}
         res=0
         last=len(strs)
         for i in strs:
             num0=i.count('0')
             num1=len(i)-num0
             temp=set()
             for j in s:
                 if j[0]+last>res and j[1]>=num0 and j[2]>=num1:
                     if j[0]+1>res:res=j[0]+1
                     temp.add((j[0]+1,j[1]-num0,j[2]-num1))
             last-=1
             s|=temp
         return res

File: Ones and Zeroes.cpp
class Solution {
 public:
     int findMaxForm(vector<string>& strs, int m, int n) {
         unordered_set<int> set;
         set.insert((m<<7)+n);
         int res=0;
         int left=strs.size();
         for(auto i:strs)
         {
             int count0=0;
             int count1=0;
             for(auto k:i)
             {
                 if(k=='0')++count0;
                 else ++count1;
             }
             unordered_set<int> tempset;
             for(auto j:set)
             {
                 int j0=(j&16760832)>>14,j1=(j&16256)>>7,j2=j&127;
                 if(j0+left>=res && j1>=count0 && j2>=count1)
                 {
                     if(res<j0+1)res=j0+1;
                     tempset.insert(((j0+1)<<14)+((j1-count0)<<7)+j2-count1);
                 }
             }
             for(auto j:tempset)
                 set.insert(j);
             --left;
         }
         return res;
     }
 };